File: To-Do-App/backend/app/schemas/task.py
# Task: T2-003 - Pydantic schemas for Task (Enhanced)
# Advanced task management schemas
"""
Pydantic schemas for Task API request/response validation.
"""

from datetime import datetime
from enum import Enum
from typing import List, Optional

from pydantic import BaseModel, Field


class Priority(str, Enum):
    """Task priority levels."""

    LOW = "low"
    MEDIUM = "medium"
    HIGH = "high"
    URGENT = "urgent"


class TaskStatus(str, Enum):
    """Task status options."""

    TODO = "todo"
    IN_PROGRESS = "in_progress"
    COMPLETED = "completed"
    CANCELLED = "cancelled"


class TaskCreate(BaseModel):
    """Schema for creating a new task."""

    title: str = Field(..., min_length=1, max_length=200)
    description: Optional[str] = Field(None, max_length=1000)
    priority: Priority = Field(default=Priority.MEDIUM)
    status: TaskStatus = Field(default=TaskStatus.TODO)
    due_date: Optional[datetime] = None
    reminder_at: Optional[datetime] = None
    start_date: Optional[datetime] = None
    category: Optional[str] = Field(None, max_length=50)
    tags: Optional[str] = Field(None, max_length=200)
    notes: Optional[str] = Field(None, max_length=2000)
    estimated_minutes: Optional[int] = Field(None, ge=0)
    is_recurring: bool = False
    recurrence_pattern: Optional[str] = Field(None, max_length=50)


class TaskUpdate(BaseModel):
    """Schema for updating an existing task."""

    title: Optional[str] = Field(None, min_length=1, max_length=200)
    description: Optional[str] = Field(None, max_length=1000)
    priority: Optional[Priority] = None
    status: Optional[TaskStatus] = None
    due_date: Optional[datetime] = None
    reminder_at: Optional[datetime] = None
    start_date: Optional[datetime] = None
    category: Optional[str] = Field(None, max_length=50)
    tags: Optional[str] = Field(None, max_length=200)
    notes: Optional[str] = Field(None, max_length=2000)
    estimated_minutes: Optional[int] = Field(None, ge=0)
    is_recurring: Optional[bool] = None
    recurrence_pattern: Optional[str] = Field(None, max_length=50)


class TaskResponse(BaseModel):
    """Schema for task response."""

    id: int
    title: str
    description: Optional[str]
    status: TaskStatus
    completed: bool
    completed_at: Optional[datetime]
    priority: Priority
    due_date: Optional[datetime]
    reminder_at: Optional[datetime]
    start_date: Optional[datetime]
    category: Optional[str]
    tags: Optional[str]
    notes: Optional[str]
    estimated_minutes: Optional[int]
    is_recurring: bool
    recurrence_pattern: Optional[str]
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


class TaskListResponse(BaseModel):
    """Schema for task list response with stats."""

    tasks: List[TaskResponse]
    total: int
    completed: int
    in_progress: int
    overdue: int
    due_today: int


class TaskStats(BaseModel):
    """Schema for task statistics."""

    total: int
    completed: int
    in_progress: int
    todo: int
    overdue: int
    due_today: int
    due_this_week: int
    by_priority: dict
    by_category: dict

File: To-Do-App/frontend/types/task.ts
// Task types with advanced features

export type Priority = "low" | "medium" | "high" | "urgent";
export type TaskStatus = "todo" | "in_progress" | "completed" | "cancelled";

export interface Task {
  id: number;
  title: string;
  description: string | null;
  status: TaskStatus;
  completed: boolean;
  completed_at: string | null;
  priority: Priority;
  due_date: string | null;
  reminder_at: string | null;
  start_date: string | null;
  category: string | null;
  tags: string | null;
  notes: string | null;
  estimated_minutes: number | null;
  is_recurring: boolean;
  recurrence_pattern: string | null;
  created_at: string;
  updated_at: string;
}

export interface TaskListResponse {
  tasks: Task[];
  total: number;
  completed: number;
  in_progress: number;
  overdue: number;
  due_today: number;
}

export interface TaskCreate {
  title: string;
  description?: string;
  priority?: Priority;
  status?: TaskStatus;
  due_date?: string;
  reminder_at?: string;
  start_date?: string;
  category?: string;
  tags?: string;
  notes?: string;
  estimated_minutes?: number;
  is_recurring?: boolean;
  recurrence_pattern?: string;
}

export interface TaskUpdate {
  title?: string;
  description?: string;
  priority?: Priority;
  status?: TaskStatus;
  due_date?: string;
  reminder_at?: string;
  start_date?: string;
  category?: string;
  tags?: string;
  notes?: string;
  estimated_minutes?: number;
  is_recurring?: boolean;
  recurrence_pattern?: string;
}

export interface TaskStats {
  total: number;
  completed: number;
  in_progress: number;
  todo: number;
  overdue: number;
  due_today: number;
  due_this_week: number;
  by_priority: Record<string, number>;
  by_category: Record<string, number>;
}

export const PRIORITY_COLORS: Record<Priority, string> = {
  low: "bg-gray-100 text-gray-700",
  medium: "bg-blue-100 text-blue-700",
  high: "bg-orange-100 text-orange-700",
  urgent: "bg-red-100 text-red-700",
};

export const STATUS_COLORS: Record<TaskStatus, string> = {
  todo: "bg-gray-100 text-gray-700",
  in_progress: "bg-yellow-100 text-yellow-700",
  completed: "bg-green-100 text-green-700",
  cancelled: "bg-red-100 text-red-700",
};

export const PRIORITY_LABELS: Record<Priority, string> = {
  low: "Low",
  medium: "Medium",
  high: "High",
  urgent: "Urgent",
};

export const STATUS_LABELS: Record<TaskStatus, string> = {
  todo: "To Do",
  in_progress: "In Progress",
  completed: "Completed",
  cancelled: "Cancelled",
};
